lib/data_hygiene/delete_topics_without_subscribers.rb

```diff
@@ -0,0 +1,106 @@
+      @input = input
+      @output = output
+    end
+
+    def call
+      puts "#{with_zero_subscribers.count} subscriber lists without subscribers"
+      with_zero_subscribers.each { |sl| puts "#{sl.gov_delivery_id} - #{sl.title}"}
+      if with_gov_delivery_error.count > 0
+        puts ''
+        puts "#{with_gov_delivery_error.count} subscriber lists with errors - THESE ARE NOT BEING DELETED"
+        with_gov_delivery_error.each { |sl| puts "#{sl.gov_delivery_id} - #{sl.title}"}
+      end
+
+      puts ''
+      puts "#{with_zero_subscribers.count} subscriber lists without subscribers, enter `delete` to delete from database and GovDelivery: "
+      command = gets
+
+      if command.chomp == 'delete'
+        with_zero_subscribers.each do |subscriber_list|
+          delete_topic(subscriber_list)
+        end
+      end
+    end
+
+    def subscriber_lists
+      @subscriber_lists ||= ThreadedSubscriberCountRetriever.new(client: client, output: @output).call
+    end
+
+    def with_zero_subscribers
+      @with_zero_subscribers ||= subscriber_lists
+        .select { |c, s| c&.zero? }
+        .map(&:last)
+    end
+
+    def with_gov_delivery_error
+      @sl_with_gov_delivery_error ||= subscriber_lists
+        .select { |c, s| c.nil? }
+        .map(&:last)
+    end
+
+    private
+
+    def delete_topic(subscriber_list)
+      topic = client.fetch_topic(subscriber_list.gov_delivery_id)
+      [topic && topic['subscribers_count'], subscriber_list]
+
+      if topic && topic['subscribers_count'] == 0
+        puts "Deleting: #{subscriber_list.gov_delivery_id}"
+        client.delete_topic(subscriber_list.gov_delivery_id)
+      elsif topic.nil?
+        puts "Skipping #{subscriber_list.gov_delivery_id} as it does not appears on GovDelivery"
+      else
+        puts "Skipping #{subscriber_list.gov_delivery_id} as it now has #{topic['subscriber_count']} subscribers"
+      end
+    end
+
+    class ThreadedSubscriberCountRetriever
+      attr_reader :client
+      delegate :puts, :print, to: :@output
+
+      def initialize(client: nil, output: STDOUT)
+        @client = client
+        @output = output
+      end
+
+      def call
+        group_size = (SubscriberList.count.to_f / thread_count).ceil
+        subscriber_list_sets = SubscriberList.all.each_slice(group_size)
+
+        results = []
+        semaphore = Mutex.new
+        threads = subscriber_list_sets.map do |subscriber_list_set|
+          Thread.new do
+            list_with_count = add_subscribers_count(subscriber_list_set)
+            semaphore.synchronize { results += list_with_count }
+          end
+        end
+        threads.each(&:join)
+        puts ''
+        results
+      end
+
+      def add_subscribers_count(subscriber_lists)
+        subscriber_lists.map do |subscriber_list|
+          print '.'
+          topic = client.fetch_topic(subscriber_list.gov_delivery_id)
+          [topic && topic['subscribers_count'], subscriber_list]
```

Comment:
Extract `fetch_topic` to its own method with error catching for know GovDelivery error codes